### PR TITLE
Val loss changes, with quite a few other changes. This is in place of the earlier PR(https://github.com/lucidrains/DALLE2-pytorch/pull/67)

### PR DESCRIPTION
**Pull request by krish240574**
This PR implements the following:

   - Validation loss is getting calculated every N steps, for a small batch of 100 embeddings
   - hyperparameters logging to wandb for each run
   - Some code cleanup in the report_cosine_sims() function.
   - load_diffusion_model and save_diffusion_model added

### CODE REVIEW

File: dalle2_pytorch/dalle2_pytorch.py [hunk 1]
```diff
@@ -23,14 +25,9 @@
 
 from resize_right import resize
 
-# rotary embeddings
```
Comment (rom1504) [inline dalle2_pytorch/dalle2_pytorch.py:26]: looks like you undid some changes here

Comment (krish240574) [reply dalle2_pytorch/dalle2_pytorch.py]: fixed, do take a look.

File: dalle2_pytorch/dalle2_pytorch.py [hunk 2]
```diff
@@ -32,6 +34,38 @@
...
+# Diffusion Prior model loading and saving functions
+
+def load_diffusion_model(dprior_path, device ):
+
+        dprior_path = Path(dprior_path)
+        assert dprior_path.exists(), 'Dprior model file does not exist'
+        loaded_obj = torch.load(str(dprior_path), map_location='cpu')
+
+        # Get hyperparameters of loaded model 
+        dpn_config = loaded_obj['hyperpmts']['Diffusion_Prior_Network']
```
Comment (rom1504) [inline dalle2_pytorch/dalle2_pytorch.py:46]: Let's call it hparams and diffusion_prior_network

File: dalle2_pytorch/dalle2_pytorch.py [hunk 3]
```diff
@@ -32,6 +34,38 @@
...
+        dp_config = loaded_obj['hyperpmts']['Diffusion_Prior']
+        image_embed_dim = loaded_obj['image_embed_dim']['image_embed_dim']
```
Comment (rom1504) [inline dalle2_pytorch/dalle2_pytorch.py:48]: let's call it diffusion_prior

File: train_diffusion_prior.py [hunk 4]
```diff
@@ -271,22 +271,40 @@ def main():
...
-      config={
-      "learning_rate": args.learning_rate,
-      "architecture": args.wandb_arch,
-      "dataset": args.wandb_dataset,
-      "epochs": args.num_epochs,
-      })
+    config = ({"learning_rate": args.learning_rate,
+        "architecture": args.wandb_arch,
+        "dataset": args.wandb_dataset,
+        "Weight Decay":args.weight_decay,
```
Comment (rom1504) [inline train_diffusion_prior.py:282]: Let's use snake case
So weight_decay
batch_size
Etc

File: dalle2_pytorch/__init__.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-from dalle2_pytorch.dalle2_pytorch import DALLE2, DiffusionPriorNetwork, DiffusionPrior, Unet, Decoder
+from dalle2_pytorch.dalle2_pytorch import DALLE2, DiffusionPriorNetwork, DiffusionPrior, Unet, Decoder,load_diffusion_model,save_diffusion_model
```
Comment (rom1504) [inline dalle2_pytorch/__init__.py:1]: Can you say in the readme these 2 methods exist?

**Comment by rom1504** [comment]
I left some comments about naming and one about doc.
The rest LGTM

Can you also take a look @lucidrains ? I think it's almost good now.

File: dalle2_pytorch/dalle2_pytorch.py (hunk 2 above)
Comment (krish240574) [reply dalle2_pytorch/dalle2_pytorch.py]: fixed

File: dalle2_pytorch/dalle2_pytorch.py (hunk 3 above)
Comment (krish240574) [reply dalle2_pytorch/dalle2_pytorch.py:48]: Fixed.

**Comment by rom1504** [comment]
Let's add some minimal doc about the load and save functions in the readme
Then LGTM

File: dalle2_pytorch/__init__.py (hunk 5 above)
Comment (krish240574) [reply dalle2_pytorch/__init__.py:1]: Fixed.

File: train_diffusion_prior.py (hunk 4 above)
Comment (krish240574) [reply train_diffusion_prior.py]: Fixed

**Comment by rom1504** [comment]
Ok LGTM

@krish240574 can you start a new run with all these last changes in? I'm curious to see how it looks on wandb

@lucidrains your turn to review :)

**Comment by krish240574** [comment]
https://wandb.ai/laion/diffusion-prior/runs/1blxu24j - wandb run for your perusal.

**Comment by rom1504** [comment]
Looks good!

**Comment by lucidrains** [comment]
@krish240574 @rom1504 thank you Kumar (and Romain for the reviewing)

this lgtm to me also :rocket: !

**Comment by lucidrains** [comment]
moving forward, i'm also going to push for an experiment tracker agnostic setup, just because of the difficulties we faced last time around trying to scale on the supercomputing cluster in Germany https://github.com/lucidrains/DALLE2-pytorch/issues/73 we can do the refactor the next PR!
